main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala

```diff
@@ -0,0 +1,243 @@
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.{LoadExpression, StoreExpression}
+import ca.uwaterloo.flix.language.{CompilationError, GenSym}
+import ca.uwaterloo.flix.language.ast.{Type, _}
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+import ca.uwaterloo.flix.util.Validation._
+
+/**
+  * A phase that merges curried definitions into definitions which take multiple
+  * arguments.
+  *
+  * For example, consider
+  *
+  *   |def add(x: Int): Int -> Int = {
+  *   |   z -> x + z
+  *   |}
+  *   |
+  *   |def a: Int = (add(3))(4)
+  *
+  * The uncurrier will create a function
+  *
+  *   |def add$uncurried(x: Int, z: Int): Int = {
+  *   |    x + z
+  *   |}
+  *
+  * And change the call site to
+  *   |def a: Int = add$uncurried(3,4)
+  */
+object Uncurrier extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+
+  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+    implicit val _ = flix.genSym
+    val startTime = System.nanoTime()
+
+    val defns = root.definitions.foldLeft(Map.empty[Symbol.DefnSym, SimplifiedAst.Definition.Constant])(Definitions.mkUncurriedDefs)
+    val uncurriedDef = defns map {case (k,v) => k -> Definitions.uncurry(v, root) }
+
+    val currentTime = System.nanoTime()
+    val time = root.time.copy(uncurrier = currentTime - startTime)
+    root.copy(definitions = uncurriedDef).copy(time = time).toSuccess
+  }
+
+  object Definitions {
+    /**
+      * Create uncurried versions of all curried functions
+      *
+      * @param map The old map of function definitions
+      * @param pair The entry which to make uncurried versions of
+      * @return A new map with the uncurried versions added
+      */
+    def mkUncurriedDefs(map: Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant], pair: (Symbol.DefnSym, SimplifiedAst.Definition.Constant))(implicit genSym: GenSym): Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant] = {
+      map ++ mkUncurriedDef(pair)
+    }
+
+    /**
+      * Creates all the uncurried versions of a function
+      *
+      * @param pair The dfnsym definition pair to base the functions on
+      * @return A map containing all the new definitions
+      */
+    def mkUncurriedDef(pair: (Symbol.DefnSym, SimplifiedAst.Definition.Constant))(implicit genSym: GenSym): Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant] = {
+      val sym = pair._1
+      val curriedDef = pair._2
+      curriedDef.exp match {
+        // If the body is a lambda, then we have a curried function
+        // definition, so we need to create an uncurried version.
+        // To do this, we'll take make the body of the function the body
+        // of the lambda, then add the arguments of the lambda to the
+        // formal parameters
+        case Lambda(args, body, _, _) =>
+          // Create a new definition for our function
+          val uncurriedSym = Symbol.mkDefnSym(pair._1.name + "$uncurried")
```

Comment:
You will need to do this (its sucks, but it is more safe)

Use `def freshDefnSym(sym: DefnSym)(implicit genSym: GenSym)` in `Symbol`.

Maintain a map, probably of this structure:

`Map[Symbol.DefnSym, Map[Int, Symbol.DefnSym]]` which given a symbol and an arity returns the appropriate uncurried function (if it exists).